File: tests/benchmarks/samples/cat-go/002_payment_processor.go
// Package payments handles charge, refund, and duplicate-charge detection
// for the storefront checkout flow.
package payments

import (
	"errors"
	"fmt"
	"log"
)

// DefaultCurrency is the currency code used when none is specified.
const DefaultCurrency = "USD"

// ErrDuplicateCharge indicates a charge was attempted twice for the same
// account within the dedupe window.
var ErrDuplicateCharge = errors.New("duplicate pending charge")

// ChargeResult holds the outcome of a successful gateway charge.
type ChargeResult struct {
	ChargeID string
	Amount   int
}

// Gateway abstracts the upstream payment provider.
type Gateway interface {
	Charge(accountID string, amount int) (ChargeResult, error)
	Refund(chargeID string) error
}

// Ledger tracks pending and settled charges per account.
type Ledger interface {
	HasPendingCharge(accountID string, amount int) bool
	MarkSettled(accountID string, chargeID string)
	MarkRefunded(chargeID string)
}

// Processor coordinates charge, refund, and duplicate-detection logic.
type Processor struct {
	gateway Gateway
	ledger  Ledger
}

// NewProcessor constructs a Processor backed by the given gateway and ledger.
func NewProcessor(gateway Gateway, ledger Ledger) *Processor {
	return &Processor{gateway: gateway, ledger: ledger}
}

// Charge charges an account, guarding against duplicate pending charges.
func (p *Processor) Charge(accountID string, amount int) (string, error) {
	if p.ledger.HasPendingCharge(accountID, amount) {
		return "", fmt.Errorf("%w: account %s", ErrDuplicateCharge, accountID)
	}
	log.Printf("charging account %s for %d", accountID, amount)
	result, err := p.gateway.Charge(accountID, amount)
	if err != nil {
		return "", err
	}
	p.ledger.MarkSettled(accountID, result.ChargeID)
	return result.ChargeID, nil
}

// Refund refunds a previously settled charge.
func (p *Processor) Refund(chargeID string) error {
	log.Printf("refunding charge %s", chargeID)
	if err := p.gateway.Refund(chargeID); err != nil {
		return err
	}
	p.ledger.MarkRefunded(chargeID)
	return nil
}

// validateAmount is a package-level helper implemented as a func literal,
// used by callers that need to check an amount before charging.
var validateAmount = func(amount int) error {
	if amount <= 0 {
		return errors.New("amount must be positive")
	}
	return nil
}

File: tests/benchmarks/samples/cat-rust/002_payment_processor.rs
//! Payment processing orchestration — charge, refund, and duplicate-charge
//! detection for the storefront checkout flow.

use std::fmt;

/// The currency code used when none is specified.
pub const DEFAULT_CURRENCY: &str = "USD";

/// Errors that can occur while charging or refunding an account.
#[derive(Debug)]
pub enum PaymentError {
    DuplicateCharge(String),
    GatewayError(String),
}

impl fmt::Display for PaymentError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            PaymentError::DuplicateCharge(account_id) => {
                write!(f, "duplicate pending charge for account {}", account_id)
            }
            PaymentError::GatewayError(message) => write!(f, "gateway error: {}", message),
        }
    }
}

/// Abstracts the upstream payment provider.
pub trait Gateway {
    fn charge(&self, account_id: &str, amount: u64) -> Result<String, PaymentError>;
    fn refund(&self, charge_id: &str) -> Result<(), PaymentError>;
}

/// Tracks pending and settled charges per account.
pub trait Ledger {
    fn has_pending_charge(&self, account_id: &str, amount: u64) -> bool;
    fn mark_settled(&mut self, account_id: &str, charge_id: &str);
    fn mark_refunded(&mut self, charge_id: &str);
}

/// Coordinates charge, refund, and duplicate-detection logic.
pub struct Processor<G: Gateway, L: Ledger> {
    gateway: G,
    ledger: L,
}

impl<G: Gateway, L: Ledger> Processor<G, L> {
    /// Construct a Processor backed by the given gateway and ledger.
    pub fn new(gateway: G, ledger: L) -> Self {
        Processor { gateway, ledger }
    }

    /// Charge an account, guarding against duplicate pending charges.
    pub fn charge(&mut self, account_id: &str, amount: u64) -> Result<String, PaymentError> {
        if self.ledger.has_pending_charge(account_id, amount) {
            return Err(PaymentError::DuplicateCharge(account_id.to_string()));
        }
        println!("charging account {} for {}", account_id, amount);
        let charge_id = self.gateway.charge(account_id, amount)?;
        self.ledger.mark_settled(account_id, &charge_id);
        Ok(charge_id)
    }

    /// Refund a previously settled charge.
    pub fn refund(&mut self, charge_id: &str) -> Result<(), PaymentError> {
        println!("refunding charge {}", charge_id);
        self.gateway.refund(charge_id)?;
        self.ledger.mark_refunded(charge_id);
        Ok(())
    }
}
